src/components/cards/TradeCardGP/TradeCardGP.vue

```diff
@@ -264,25 +275,83 @@ export default defineComponent({
     function onSuccessOverlayClose() {
       tradeSuccess.value = false;
       orderMetadata.value = null;
+      txHash.value = '';
+      orderId.value = '';
     }
 
     async function trade() {
       try {
-        const unsignedOrder: UnsignedOrder = {
-          sellToken: normalizeTokenAddress(tokenInAddress.value),
-          buyToken: normalizeTokenAddress(tokenOutAddress.value),
-          sellAmount: parseUnits(
-            tokenInAmount.value,
+        const { chainId } = getConfig();
+        const tokenInAmountScaled = scale(
+          bnum(tokenInAmount.value),
+          tokenInDecimals.value
+        );
+
+        const tokenOutAmountScaled = scale(
+          bnum(tokenOutAmount.value),
+          tokenOutDecimals.value
+        );
+
+        trading.value = true;
+
+        if (isWrap.value) {
+          try {
+            const tx = await wrap(chainId, auth.web3, tokenInAmountScaled);
+            console.log('Wrap tx', tx);
+            tradeTxListener(tx.hash);
+          } catch (e) {
+            console.log(e);
+            trading.value = false;
+          }
+          return;
+        } else if (isUnwrap.value) {
+          try {
+            const tx = await unwrap(chainId, auth.web3, tokenInAmountScaled);
+            console.log('Unwrap tx', tx);
+            tradeTxListener(tx.hash);
+          } catch (e) {
+            console.log(e);
+            trading.value = false;
+          }
+          return;
+        }
+
+        let sellAmount: string;
+        let buyAmount: string;
+
+        if (exactIn.value) {
+          sellAmount = parseUnits(tokenInAmount.value, tokenInDecimals.value)
+            .sub(feeAmount.value)
+            .toString();
+          buyAmount = parseUnits(
+            tokenOutAmountScaled
+              .div(1 + slippageBufferRate.value)
+              .integerValue(BigNumber.ROUND_DOWN)
+              .toString(),
+            tokenOutDecimals.value
+          ).toString();
+        } else {
+          sellAmount = parseUnits(
+            tokenInAmountScaled
+              .times(1 + slippageBufferRate.value)
+              .integerValue(BigNumber.ROUND_DOWN)
+              .toString(),
```

Comment:
I'm not a fan of these massive functions. Could we not break it down a little so it's more readable? e.g.
```ts
async function trade() {
  if (isWrap.value) {
    await wrap();
  } else if (isUnwrap.value) {
    await unwrap();
  } else {
    await executeTrade();
  }
}
```
Doesn't have to be like this exactly or named like this, I just think it could be broken down a bit.

Reply:
Yeah totally agree - I split the functions.

The code is likely to keep changing a lot tho, so this isn't the final one! Will try to keep functions smaller and more readable.